### ISSUE TITLE
linkchecker failing on UTF-8 filenames - but only in subfolder with index.html file

### ISSUE THREAD

**Issue by matkoniecz** [linkchecker/linkchecker#554]
## Summary

Filename containing nonASCII "ń" ( https://en.wikipedia.org/wiki/Ogonek ) is sometimes incorrectly detected as unlinked

I prepared bug reproduction testcase in https://github.com/matkoniecz/linkchecker_utf8_bug

## Steps to reproduce

 1. `git clone https://github.com/matkoniecz/linkchecker_utf8_bug.git`
 2. `cd linkchecker_utf8_bug`
 3. `linkchecker index.html`

## Actual result

Incorrect errors are reported

## Expected result

No errors are reported
## Environment
<!-- replace the comments with the output of the commands -->

 * Operating system:

```
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal
```

 * Linkchecker version:

```
LinkChecker 10.0.1 released 29.1.2021
Copyright (C) 2000-2016 Bastian Kleineidam, 2010-2021 LinkChecker Authors
```

 * Python version: 

```
Python 3.8.10
```

 * Install method: from git as requested in readme
 * Site URL: local file

## Configuration file

None

`cat: /home/mateusz/.linkchecker/linkcheckerrc: No such file or directory`

## Logs

```
linkchecker --verbose index.html 
INFO linkcheck.cmdline 2021-08-14 15:32:56,027 MainThread Checking intern URLs only; use --check-extern to check extern URLs.
LinkChecker 10.0.1
Copyright (C) 2000-2016 Bastian Kleineidam, 2010-2021 LinkChecker Authors
LinkChecker comes with ABSOLUTELY NO WARRANTY!
This is free software, and you are welcome to redistribute it under
certain conditions. Look at the file `LICENSE' within this distribution.
Get the newest version at https://linkchecker.github.io/linkchecker/
Write comments and bugs to https://github.com/linkchecker/linkchecker/issues

Start checking at 2021-08-14 15:32:56+002

URL        `osm_to_svg_in_browser/laser_neighbourhood_-_Madali%C3%85%E2%80%9Eskiego.png'
Parent URL file:///home/mateusz/Desktop/tmp/linkchecker_utf8_bug/index.html, line 4, col 5
Real URL   file:///home/mateusz/Desktop/tmp/linkchecker_utf8_bug/osm_to_svg_in_browser/laser_neighbourhood_-_Madali%C3%83%E2%80%A6%C3%A2%E2%82%AC%C5%BEskiego.png
Check time 0.000 seconds
Result     Error: URLError: <urlopen error [Errno 2] No such file or directory: '/home/mateusz/Desktop/tmp/linkchecker_utf8_bug/osm_to_svg_in_browser/laser_neighbourhood_-_MadaliÃ…â€žskiego.png'>

URL        `file:///home/mateusz/Desktop/tmp/linkchecker_utf8_bug/index.html'
Name       `index.html'
Real URL   file:///home/mateusz/Desktop/tmp/linkchecker_utf8_bug/index.html
Check time 0.001 seconds
Size       294B
Modified   2021-08-14 13:32:45.296032Z
Result     Valid

URL        `osm_to_svg_in_browser/index.html'
Name       `tool that I made'
Parent URL file:///home/mateusz/Desktop/tmp/linkchecker_utf8_bug/index.html, line 6, col 18
Real URL   file:///home/mateusz/Desktop/tmp/linkchecker_utf8_bug/osm_to_svg_in_browser/index.html
Check time 0.001 seconds
Size       125B                                                                                                                                                                              
Modified   2021-08-14 13:32:45.296032Z                                                                                                                                                       
Result     Valid                                                                                                                                                                             
                                                                                                                                                                                             
Statistics:                                                                                                                                                                                  
Downloaded: 419B.                                                                                                                                                                            
Content types: 0 image, 2 text, 0 video, 0 audio, 0 application, 0 mail and 1 other.                                                                                                         
URL lengths: min=64, max=150, avg=100.

That's it. 3 links in 3 URLs checked. 0 warnings found. 1 error found.
Stopped checking at 2021-08-14 15:32:57+002 (1 seconds)
```

## Other notes

`linkchecker --verbose https://matkoniecz.github.io/linkchecker_utf8_bug/` is not buggy

Note that the same file in index.html is not triggering problems

Note that https://matkoniecz.github.io/linkchecker_utf8_bug/ works fine

**Comment by matkoniecz** [comment]
I was investigating it further and it appears to be broken only for index.html files. See https://github.com/matkoniecz/linkchecker_utf8_bug/commit/6070ead13ed9d2255a4f1f8108f392f6ab8bb62c

**Comment by cjmayo** [comment]
I've identified a LinkChecker problem and a solution in #572 for http(s) links. Testing welcome. I think that is all LinkChecker can do - it may not be sufficient.

Even without that commit, checking the page as a local file from linkchecker (i.e. not via http) can help debug your system. If it still doesn't get the encoding right, do you have chardet installed? Which version?

Then one further option:
- install [cchardet](https://pypi.org/project/cchardet/) #571

Another:
- add `<meta charset="UTF-8">` to the html

**Comment by tgoeg** [comment]
Alright, this took me the full day today to hunt this down.

In my case, this seems to be a combination of a bug both in Apache and in Linkchecker (but please correct me if I'm wrong).

It all starts with Apache 2.4.41 on a bog standard Ubuntu 20.04 LTS responding to a request for domain.com/path with a redirect to domain.com/path/ (just serving static pages at /index.htm and /path/index.htm etc.). This also seems to be what @matkoniecz sees when he says it only occurs in subfolders!

However, Apache sets
```Content-Type text/html; charset=iso8859-1```
even if I set
```AddDefaultCharset UTF-8```
in the vhost config (per vhost or per directory, does not matter). This only seems to apply to regularly delivered pages, not non-2xx ones. I consider this a bug in Apache, as https://httpd.apache.org/docs/current/mod/mod_headers.html#Header implicitly states that setting (well, at least "adding") headers for non-2xx pages should be possible:

> "[..]You're adding a header to a locally generated non-success (non-2xx) response, such as a redirect, in which case only the table corresponding to always is used in the ultimate response.[..]"

The actual target page at /path/ gets sent out correctly with `charset=utf.8` and this is the part where I think linkchecker does something weird, as it seems to take the charset of the first page redirecting it (iso8859-1) and applies it to the destination page (utf-8). Running linkchecker with `-v` also shows that the "direct" url at /path/ is checked earlier, and correctly, and then there's another check (for the now "different"=wrongly encoded url) that fails.
I guess this needs to be fixed in Linkchecker.

This is some kind of regression, as Linkchecker 9.4.0 does not face this problem.

Workaround for the time being, if you are in control of the server and if you know your page only has utf-8 content:
Add
```Header always edit Content-Type iso-8859-1 utf-8```
to your vhost config. Strangely, this fixes 301 pages as well.

This at least fixes it for me and I'd be very glad if anyone could confirm my findings (or whether I found something completely unreleated ;-) )

**Comment by matkoniecz** [comment]
> Workaround for the time being, if you are in control of the server and if you know your page only has utf-8 content:

At least for now all my hosting is handled by github pages (due to combination of price, ease of use and that it fulfills nearly all of my needs)

**Comment by cjmayo** [comment]
Still need to spend some time (hopefully less than a day...) to understand the subfolder part. N.B. The PR will also print the server response headers, when using `-D checking`.

**Comment by cjmayo** [comment]
I've tried to reproduce with a subdirectory but I still come back to chardet.

```py
>>> chardet.detect_all("ń".encode("utf-8"))
[{'encoding': 'Windows-1252', 'confidence': 0.73, 'language': ''}, {'encoding': 'utf-8', 'confidence': 0.505, 'language': ''}]

>>> cchardet.detect("ń".encode("utf-8"))
{'encoding': 'UTF-8', 'confidence': 0.5049999952316284}
```

@matkoniecz do you have chardet installed (and not cchardet)?

If so I suggest installing cchardet (or for this test anyway, uninstalling chardet) and rerunning your file:// test.

**Comment by matkoniecz** [comment]
Neither is installed

```
mateusz@grima:~/Documents/install_moje/OSM software/leafleter_python$ chardet

Command 'chardet' not found, but can be installed with:

sudo apt install python-chardet

mateusz@grima:~/Documents/install_moje/OSM software/leafleter_python$ cchardet

Command 'cchardet' not found, did you mean:

  command 'chardet' from deb python-chardet (3.0.4-4build1)
  command 'uchardet' from deb uchardet (0.0.6-3build1)

Try: sudo apt install <deb name>

```

**Comment by cjmayo** [comment]
No. I meant https://pypi.org/project/cchardet/. Don't think it is available in Debian/Ubuntu. So that would have to be a `pip3` installation.

Interesting though, I tried without both and it worked for me.

**Comment by cjmayo** [comment]
chardet would be python3-chardet though? Sure you don't have that?

**Comment by matkoniecz** [comment]
```
mateusz@grima:~/Documents/install_moje/OSM software/leafleter_python$ pip3 install chardet
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: chardet in /usr/lib/python3/dist-packages (3.0.4)
mateusz@grima:~/Documents/install_moje/OSM software/leafleter_python$ pip3 install cchardet
Defaulting to user installation because normal site-packages is not writeable
Collecting cchardet
  Downloading cchardet-2.1.7-cp38-cp38-manylinux2010_x86_64.whl (265 kB)
     |████████████████████████████████| 265 kB 1.5 MB/s            
Installing collected packages: cchardet
Successfully installed cchardet-2.1.7
```

**Comment by matkoniecz** [comment]
OK, I now tested it: with `pip3 install cchardet` it works without false positives.

After `cchardet` is removed and only `chardet` is present I get false positive.

**Comment by cjmayo** [comment]
Great - we're seeing the same thing.
Looks like you always had python3-chardet installed (which explains what you saw initially), because pip actually skipped installing chardet:
```
Requirement already satisfied: chardet in /usr/lib/python3/dist-packages (3.0.4)
```

Thanks for reporting and testing. As a result we now have an explanation in install.txt.

**Comment by tgoeg** [comment]
python3-chardet is also installed on my system.
That's because the debian linkchecker pkg pulls it in:

```
~# apt-cache depends linkchecker
linkchecker
  Depends: python3
  Depends: python3-bs4
  Depends: python3-requests
  Depends: python3-urllib3
  Depends: python3-xdg
  Depends: python3-dnspython
  Depends: <python3:any>
    python3:i386
    python3
  Suggests: clamav-daemon
  Suggests: <linkchecker-web>
  Suggests: python3-argcomplete
  Suggests: python3-cssutils
  Suggests: <python3-gconf>
  Suggests: python3-geoip
  Suggests: <python3-meliae>
```
and
```
# apt-cache depends python3-requests
python3-requests
  Depends: python3-certifi
  Depends: python3-chardet
  Depends: python3-idna
  Depends: python3-urllib3
  Depends: <python3:any>
    python3:i386
    python3
  Depends: ca-certificates
  Depends: python3-chardet
  Depends: python3-urllib3
  Breaks: awscli
  Suggests: python3-cryptography
  Suggests: python3-idna
  Suggests: python3-openssl
  Suggests: python3-socks
```
Shouldn't this be filed against the debian pkg, then? There's no python3-cchardet pkg however..

**Comment by matkoniecz** [comment]
@tgoeg Once `apt` packages new version that includes https://github.com/linkchecker/linkchecker/pull/571/files they will presumably change dependencies and will need to package `python3-cchardet`

**Comment by tgoeg** [comment]
Ah, alright, so this *can* be solved right now by (optionally) installing cchardet along 10.0.1 and it *will* get solved with a new release that has this mandatory dep.
Thanks for the heads up.

**Comment by cjmayo** [comment]
Gosh, it's complicated...

I've merged the PR for http servers returning no charset in Content-Types.

But, I've gone a bit cautious on only recommending cchardet. PR #571 updated to try and explain the whole situation instead.

Looks like on Debian/Ubuntu it is python3-bs4 that is recommending (not requiring python3-chardet).

I've tried uninstalling both chardet and cchardet with some success (setting up a venv and just installing linkchecker from git+https), just relying on Beautiful Soup.

N.B. the new Docker images contain cchardet.

**Comment by tgoeg** [comment]
Not sure if I understand you right or you probably missed that:
I'm on Debian and Ubuntu and python3-**requests** has a **hard dep** on python3-chardet (see https://github.com/linkchecker/linkchecker/issues/554#issuecomment-981714309) and not just a recommend. So if linkchecker needs python3-requests (which it currently does, in the .deb pkg version at least), it will pull in python3-chardet no matter what.

Additionally, in https://github.com/linkchecker/linkchecker/commit/cc911d6621edfa22f3b6b1e7b81dc681332c8b0e you linked, I'd definitely add the issue I described above:
First, Apache ignores the default charset for error pages/redirects (and people might think their charsets are correctly setup if they don't dig for hours and sift through linkchecker's debug log with the linked patch applied)
Second, linkchecker uses that wrong charset for the resulting page it gets redirected to (which actually has the correct charset; I still consider this to be a bug in linkchecker or its dependencies - shall I open a separate issue for it?)

**Comment by matkoniecz** [comment]
> So if linkchecker needs python3-requests (which it currently does, in the .deb pkg version at least), it will pull in python3-chardet no matter what.

That is not a problem. If cchardet will be also present then it will be used.

And anyway is Debian packaging even on topic here?

**Comment by cjmayo** [comment]
> Not sure if I understand you right or you probably missed that: I'm on Debian and Ubuntu and python3-**requests** has a **hard dep** on python3-chardet (see [#554 (comment)](https://github.com/linkchecker/linkchecker/issues/554#issuecomment-981714309)) and not just a recommend. So if linkchecker needs python3-requests (which it currently does, in the .deb pkg version at least), it will pull in python3-chardet no matter what.

OK. I'm using Requests 2.26.0 which will use charset_normalizer (if chardet is not installed...)
https://docs.python-requests.org/en/latest/community/updates/#id1

I will write some more on chardet/cchardet/bs4 another time.

> 
> Additionally, in [cc911d6](https://github.com/linkchecker/linkchecker/commit/cc911d6621edfa22f3b6b1e7b81dc681332c8b0e) you linked, I'd definitely add the issue I described above: First, Apache ignores the default charset for error pages/redirects (and people might think their charsets are correctly setup if they don't dig for hours and sift through linkchecker's debug log with the linked patch applied) Second, linkchecker uses that wrong charset for the resulting page it gets redirected to (which actually has the correct charset; I still consider this to be a bug in linkchecker or its dependencies - shall I open a separate issue for it?)

First I've had a guess (I'm unable to test) #579 at a solution. If you are able to test then let's take the redirect discussion to that PR for now. Else, yes I think a new issue would be the way to go.

**Comment by cjmayo** [comment]
OK, so here's what I think the story is (do correct me - I'm also going to ignore the redirects for brevity, do see 579):

1) UTF-8 page served with a Content-Type that did not include charset=UTF-8
2) Requests reported the encoding as ISO-8859-1
3) LinkChecker used the ISO-8859-1 encoding and mangled the page
4) Minimal example created to reproduce the problem. Usually absolutely the right thing to do but...
5) Because chardet was already installed it was used by Beautiful Soup to detect the encoding of the file
6) chardet is not good with short strings https://github.com/chardet/chardet/issues/242#issuecomment-914539641
7) cchardet is better at short strings
8) Beautiful Soup might be enough on its own, but if it has access to chardet it will use it without asking
9) If it has access to cchardet it will use that as its first preference

The Content-Type issue has fixed in #572.
Proposed some words for install.txt in #571 to try and make people aware of the chardet/cchardet/bs4 situation - I've just added a bit about using a venv.

I'm still unclear how chardet performs in the real world on HTML.

**Comment by tgoeg** [comment]
I know you intentionally wanted to leave out the redirect problem, but I think this very issue is more or less only the redirect problem itself, also in @matkoniecz's case (as the "only in subfolder" case causes the redirect, at least at my site)
I cannot confirm whether your rundown actually applies in other cases than the redirection problem, but probably, yes.
It does not really matter whether the wrong iso8859-1 charset is detected from the redirect before or from the headers of the actual page, so starting at 3. it should be the same, yes.

However, I don't see any issue like 6., so I cannot confirm. As soon as the wrong iso8859-1 charset carried over from the redirect was corrected, the actual charset detection worked our perfectly in my case with default chardet from Debian packages, I did not have to change anything more than the diff in #579's PR

This might however not be the case in other (corner) cases/open issues I don't know of!

It might probably help if @matkoniecz could confirm whether #579 helps in this very issue's case as well.

I'd also add a note regarding the Apache issue/bug(?) as people will think they do have their server setup correctly as stated in the readme. I filed a bug already some time ago, but there's no answer yet: https://bz.apache.org/bugzilla/show_bug.cgi?id=65699

**Comment by cjmayo** [comment]
One extra thing, the Requests behaviour is documented:
https://docs.python-requests.org/en/latest/user/advanced/#encodings
and follows RFC 2616.

**Comment by cjmayo** [comment]
Updated install.txt and redirect encoding setting merged.

I suggest the [Wiki](https://github.com/linkchecker/linkchecker/wiki) would be the best place for topics like Apache that aren't part of LinkChecker but are of interest to users. Plus no need for PRs and easy to update.

Do please report any problems on new issues.

**Comment by tgoeg** [comment]
Added a short article in the wiki.
